File: clients/fsharp/DebugInterface.fs
namespace Komus24

type DebugInterface(reader, writer) =
    member this.addCircle(pos: Model.Vec2Double, radius: double, color: Debugging.Color) =
        this.add (Debugging.DebugData.Circle {
            Pos = pos
            Radius = radius
            Color = color
        })
    
    member this.addLine(point1: Model.Vec2Double, point2: Model.Vec2Double, width: double, color: Debugging.Color) =
        this.add (Debugging.DebugData.Line {
            Point1 = point1
            Point2 = point2
            Width = width
            Color = color
        })
    
    member this.addRect(corner1: Model.Vec2Double, corner2: Model.Vec2Double, color: Debugging.Color) =
        this.add (Debugging.DebugData.Rect {
            Corner1 = corner1
            Corner2 = corner2
            Color = color
        })
    
    member this.addText(text: string, pos: Model.Vec2Double, size: double, align: double, color: Debugging.Color) =
        this.add (Debugging.DebugData.Text {
            Text = text
            Pos = pos
            Size = size
            Align = align
            Color = color
        })

    member this.add(debugData: Debugging.DebugData) =
        this.send (Debugging.DebugCommand.Add {
            DebugData = debugData
        })
    
    member this.clear() =
        this.send (Debugging.DebugCommand.Clear (new Debugging.DebugCommandClear()))
    
    member this.setAutoFlush(enable: bool) =
        this.send (Debugging.DebugCommand.SetAutoFlush {
            Enable = enable
        })
    
    member this.flush() =
        this.send (Debugging.DebugCommand.Flush (new Debugging.DebugCommandFlush()))

    member this.send(command: Debugging.DebugCommand) =
        (Codegame.ClientMessage.DebugMessage { Command = command }).writeTo writer
        writer.Flush()

    member this.getState(): Debugging.DebugState =
        (new Codegame.ClientMessageRequestDebugState()).writeTo writer
        writer.Flush()
        Debugging.DebugState.readFrom reader

File: clients/rust/src/debug_interface.rs
use komus24::*;

pub struct DebugInterface<'a> {
    reader: &'a mut dyn std::io::Read,
    writer: &'a mut dyn std::io::Write,
}

impl<'a> DebugInterface<'a> {
    pub fn new(reader: &'a mut dyn std::io::Read, writer: &'a mut dyn std::io::Write) -> Self {
        Self { reader, writer }
    }

    pub fn add_circle(&mut self, pos: model::Vec2F64, radius: f64, color: debugging::Color) {
        self.add(debugging::DebugData::Circle { pos, radius, color });
    }
    
    pub fn add_line(&mut self, point1: model::Vec2F64, point2: model::Vec2F64, width: f64, color: debugging::Color) {
        self.add(debugging::DebugData::Line { point1, point2, width, color });
    }
    
    pub fn add_rect(&mut self, corner1: model::Vec2F64, corner2: model::Vec2F64, color: debugging::Color) {
        self.add(debugging::DebugData::Rect { corner1, corner2, color });
    }
    
    pub fn add_text(&mut self, text: String, pos: model::Vec2F64, size: f64, align: f64, color: debugging::Color) {
        self.add(debugging::DebugData::Text { text, pos, size, align, color });
    }

    pub fn add(&mut self, debug_data: debugging::DebugData) {
        self.send(debugging::DebugCommand::Add { debug_data });
    }
    
    pub fn clear(&mut self) {
        self.send(debugging::DebugCommand::Clear {  });
    }
    
    pub fn set_auto_flush(&mut self, enable: bool) {
        self.send(debugging::DebugCommand::SetAutoFlush { enable });
    }
    
    pub fn flush(&mut self) {
        self.send(debugging::DebugCommand::Flush {  });
    }

    pub fn send(&mut self, command: debugging::DebugCommand) {
        use trans::Trans;
        codegame::ClientMessage::DebugMessage { command }
            .write_to(self.writer)
            .expect("Failed to write custom debug data");
        self.writer.flush().expect("Failed to flush");
    }

    pub fn get_state(&mut self) -> debugging::DebugState {
        use trans::Trans;
        codegame::ClientMessage::RequestDebugState {}
            .write_to(self.writer)
            .expect("Failed to write request debug state message");
        self.writer.flush().expect("Failed to flush");
        debugging::DebugState::read_from(self.reader).expect("Failed to read debug state")
    }
}